### ISSUE TITLE
FillMaskPipeline assumes model return dict but return_dict is not set

### ISSUE THREAD

**Issue by richardwth** [huggingface/transformers#14505]
## Environment info

- `transformers` version: 4.11.3
- Platform: Linux
- Python version: 3.6.13
- PyTorch version (GPU?): 1.9
- Tensorflow version (GPU?): N/A
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

### Who can help 
@Narsil

## Information

Model I am using (Bert, XLNet ...): Bert, Roberta, Deberta

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

I came across this error 
`'tuple' object does not support item assignment` 
when using `nlpaug` package ([link](https://github.com/makcedward/nlpaug)) for data augmentation. The code snippet is
```Python
import nlpaug.augmenter.word as naw
text = 'The chef cook the meal.'
aug = naw.ContextualWordEmbsAug(
    model_path=model_path, 
    model_type='roberta',
    action="insert")
augmented_text = aug.augment(text)
```
It is easy to locate this bug by checking `transformers.pipelines.fill_mask.py` [line 90](https://github.com/huggingface/transformers/blob/956a483173e77ebf655ca9636a5f7b6ef010b307/src/transformers/pipelines/fill_mask.py#L90):
```Python
def _forward(self, model_inputs):
        model_outputs = self.model(**model_inputs)
        model_outputs["input_ids"] = model_inputs["input_ids"]
        return model_outputs
```
and [line 184](https://github.com/huggingface/transformers/blob/956a483173e77ebf655ca9636a5f7b6ef010b307/src/transformers/pipelines/fill_mask.py#L184):
```Python
def _sanitize_parameters(self, top_k=None, targets=None):
        postprocess_params = {}

        if targets is not None:
            target_ids = self.get_target_ids(targets, top_k)
            postprocess_params["target_ids"] = target_ids

        if top_k is not None:
            postprocess_params["top_k"] = top_k

        if self.tokenizer.mask_token_id is None:
            raise PipelineException(
                "fill-mask", self.model.base_model_prefix, "The tokenizer does not define a `mask_token`."
            )
        return {}, {}, postprocess_params
```
The `_forward` assumes `model_outputs` is a `dict` but models like Bert, Roberta assume `return_dict=None` by default; the `_sanitize_parameters` returns empty `forward_params` so `model_inputs` does not contain `{'return_dict': True}`, thus causing the error.

## Expected behavior

Please do either
- Allow `_sanitize_parameters` to accept `**kwargs` and parse/keep `{'return_dict': True}` as `forward_params`; allow `_forward` to accept `**forward_params` and pass them to model. 
- Explicitly set `model_inputs['return_dict'] = True` in `_forward` (if other models also use this `return_dict` argument to decide whether the output is tuple or dict).
- Provide other suggestions.

**Comment by Narsil** [comment]
Ok, `_forward` assumes that the model will returns a `BaseModelOutput` which are Transformers specific dicts which are necessary to support all backends (PT, TF, flax).

There won't be support for tuples since it's harder and less readable to use.
The pipelines do need to change what get passed on to `postprocess` since information from `preprocess` might still be important (here `input_ids` for instance).
In addition, the pipeline can handle batching and auto padding which is impossible to handle without knowing tensor names. (since the padding value depends on the type of tensor)

AFAIK there is not downside to using `BaseModelOutput`. In general, `return_dict` does not have to be set, since it's the default of the library.

I am not sure where this information is lost in your script since you share only a very small amount of code.
Could you share a reproducible example so I can probably provide a better workaround ?

**Comment by richardwth** [comment]
> Ok, `_forward` assumes that the model will returns a `BaseModelOutput` which are Transformers specific dicts which are necessary to support all backends (PT, TF, flax).
> 
> There won't be support for tuples since it's harder and less readable to use. The pipelines do need to change what get passed on to `postprocess` since information from `preprocess` might still be important (here `input_ids` for instance). In addition, the pipeline can handle batching and auto padding which is impossible to handle without knowing tensor names. (since the padding value depends on the type of tensor)
> 
> AFAIK there is not downside to using `BaseModelOutput`. In general, `return_dict` does not have to be set, since it's the default of the library.
> 
> I am not sure where this information is lost in your script since you share only a very small amount of code. Could you share a reproducible example so I can probably provide a better workaround ?

Thanks for your reply and sorry for a waste of your time.

I realize the issue was on my side that `return_dict` was set to False in the provided checkpoint at `model_path`. The default of the library is indeed to return a dictionary.

**Comment by LysandreJik** [comment]
I think forcing `return_dict` to be `True` has no downsides and would prevent issues where a model has `return_dict` set to `False` before being loaded in the pipeline. I believe we would be open for an explicit setting of `return_dict` to `True` in the pipelines.

**Comment by Narsil** [comment]
@LysandreJik do all models in all frameworks support the kwarg ?

If it's not an overridable option, we could definitely change every call to force `return_dict=True`

```python
self.model(**model_inputs)
```

to 
```python
self.model(**model_inputs, return_dict=True)
```

Would that work ?
